src/Control/FunFlow/ContentStore.hs

```diff
@@ -0,0 +1,289 @@
+-- externally. The contents of subtrees may be modified externally while the
+-- subtree is marked as under construction.
+--
+-- __Implementation note:__
+--
+-- Two file-system features are used to persist the state of a subtree,
+-- namely whether it exists and whether it is writable.
+--
+-- @
+--   exists   writable          state
+--   ---------------------------------------
+--                             missing
+--     X          X       under construction
+--     X                       complete
+-- @
+module Control.FunFlow.ContentStore
+  ( Status (..)
+  , StoreError (..)
+  , ContentStore
+  , root
+  , initialize
+  , allSubtrees
+  , subtrees
+  , subtreesUnderConstruction
+  , query
+  , isMissing
+  , isUnderConstruction
+  , isComplete
+  , lookup
+  , markUnderConstruction
+  , markComplete
+  , removeFailed
+  , removeForcibly
+  ) where
+
+
+import           Prelude                         hiding (lookup)
+
+import           Control.Concurrent.MVar
+import           Control.Exception               (Exception, bracket_, throwIO)
+import           Control.Monad                   (filterM)
+import           Data.Bits                       (complement)
+import           Data.List                       (foldl')
+import           Data.Maybe                      (catMaybes)
+import           Data.Typeable                   (Typeable)
+import           System.Directory                (createDirectory,
+                                                  createDirectoryIfMissing,
+                                                  doesDirectoryExist,
+                                                  listDirectory, makeAbsolute,
+                                                  removePathForcibly)
+import           System.FilePath                 ((</>))
+import           System.Posix.Files
+import           System.Posix.Types
+
+import           Control.FunFlow.ContentHashable (ContentHash, hashToPath,
+                                                  pathToHash)
+
+
+-- | Status of a subtree in the store.
+data Status
+  = Missing
+  -- ^ The subtree does not exist, yet.
+  | UnderConstruction
+  -- ^ The subtree is under construction and not ready for consumption.
+  | Complete
+  -- ^ The subtree is complete and ready for consumption.
+  deriving (Eq, Show)
+
+-- | Errors that can occur when interacting with the store.
+data StoreError
+  = NotUnderConstruction ContentHash
+  -- ^ A subtree is not under construction when it should be.
+  | AlreadyUnderConstruction ContentHash
+  -- ^ A subtree is already under construction when it should be missing.
+  | AlreadyComplete ContentHash
+  -- ^ A subtree is already complete when it shouldn't be.
+  deriving (Show, Typeable)
+instance Exception StoreError
+
+-- | A hash addressed store on the file system.
+data ContentStore = ContentStore
+  { storeRoot :: FilePath
+  -- ^ Subtrees are stored directly under this directory.
+  , storeLock :: MVar ()
```

Comment:
Should clarify that this is a write lock?

Reply:
Yeah, I should clarify that. It's a metadata lock. Multiple processes can write into multiple subtrees under construction simultaneously, but marking a subtree as under construction, or ready, or removing it, takes the lock.